### ISSUE TITLE
Computation with distributed arrays results in unexpected PartitionSpec

### ISSUE THREAD

**Issue by shoyer** [google/jax#15782]
### Description

Computation with distributed arrays seems to normalize `PartitionSpec` in at least three ways:

- Leading axes of length 1 are replaced by `None`.
- Trailing axes of length 1 are removed.
- Strings are replaced by tuples of strings.

I found this behavior rather unexpected, because this normalization is applied and validated somewhat inconsistently:
1. `PartitionSpec` objects don't get normalized when put into a `NamedSharding`.
1. Partition specs on arrays created with `device_put` also are not get normalized.
2. Differently normalized `NamedSharding` objects sometimes (but not always) compare equal.
3. Differently normalized `PartitionSpec` objects never compare equal.

Example code:
```python
import chex
import jax
import jax.numpy as jnp
import numpy as np

chex.set_n_cpu_devices(8)
P = jax.sharding.PartitionSpec

mesh = jax.sharding.Mesh(jax.devices(), axis_names=['x'])
spec = P('x')
sharding = jax.sharding.NamedSharding(mesh, spec)

x = jax.device_put(jnp.arange(32), sharding)
y = x + 1

print('example one:')
print(f'{x.sharding=}')
print(f'{y.sharding=}')
print(x.sharding == y.sharding)
print(x.sharding.spec == y.sharding.spec)

devices = np.reshape(jax.devices(), (1, 8, 1))
mesh = jax.sharding.Mesh(devices, axis_names=['x', 'y', 'z'])
spec = P('x', 'y', 'z')
sharding = jax.sharding.NamedSharding(mesh, spec)

x = jax.device_put(jnp.arange(32).reshape(1, -1, 1), sharding)
y = x + 1

print()
print('example two:')
print(f'{x.sharding=}')
print(f'{y.sharding=}')
print(x.sharding == y.sharding)
print(x.sharding.spec == y.sharding.spec)
```
Prints:
```
example one:
x.sharding=NamedSharding(mesh={'x': 8}, spec=PartitionSpec('x',))
y.sharding=NamedSharding(mesh={'x': 8}, spec=PartitionSpec(('x',),))
True
False

example two:
x.sharding=NamedSharding(mesh={'x': 1, 'y': 8, 'z': 1}, spec=PartitionSpec('x', 'y', 'z'))
y.sharding=NamedSharding(mesh={'x': 1, 'y': 8, 'z': 1}, spec=PartitionSpec(None, ('y',)))
False
False
```

**Comment by shoyer** [comment]
I think normalization is fine, but it's less suprising to users if it's done consistently, e.g., JAX should either never normalize, or normalize in every case where it's possible.

As a concrete suggestion, normalizing strings to length 1 tuples of strings could be done every time a `PartitionSpec` is created, and removing length 1 dimensions could be done every time a partition spec is put inside a `NamedSharding`.

> If you want to compare 2 shardings, you can use `==` or `is_equivalent_to` for comparison which will do all the necessary canonicalizations for you.

As noted above (see my "example two"), this is currently not true. Equivalent `NamedSharding` objects do not always compare as equal. I think this is probably a bug? My guess is that it may result in some unexpected cache misses.

**Comment by yashk2810** [comment]
> Equivalent NamedSharding objects do not always compare as equal.

They do. The equality is not implemented by comparing the PartitionSpecs. Try doing `x.sharding == y.sharding` instead of accessing `spec`.

**Comment by yashk2810** [comment]
Ohh sorry.. I missed that you were doing that. Hmm.. let me check

**Comment by yashk2810** [comment]
I think this is WAI. There won't be any cache misses though. You can test that by running it via jit and check that the compilation cache is hit.

**Comment by yashk2810** [comment]
I can probably fix the tuple thing for you though. That is a reasonable request. Let me send a CL for it.

Fun fact, I tried to fix it almost 1 year ago to normalize everything but don't remember why I never submitted that CL.

**Comment by shoyer** [comment]
Building on my second example, I do see a cache miss if you try to pass an equivalent sharding that does not compare as equal as a static argument to jit:
```
import chex
import jax
import jax.numpy as jnp
import numpy as np
import functools

@functools.partial(jax.jit, static_argnames=['sharding'])
def f(x, sharding):
  print('compiling')
  return x

chex.set_n_cpu_devices(8)
P = jax.sharding.PartitionSpec

devices = np.reshape(jax.devices(), (1, 8, 1))
mesh = jax.sharding.Mesh(devices, axis_names=['x', 'y', 'z'])
spec = P('x', 'y', 'z')
sharding = jax.sharding.NamedSharding(mesh, spec)

x = jax.device_put(jnp.arange(32).reshape(1, -1, 1), sharding)
y = x + 1

f(x, x.sharding)  # prints "compiling"
f(x, y.sharding)  # prints "compiling" again
```

**Comment by yashk2810** [comment]
Good point. About the cache miss, I meant that without the static_argnames you won't see a cache_miss.

I'll try to fix this (but I don't guarantee anything :)). Thanks for the repro!

**Comment by shoyer** [comment]
I would definitely fix the equality comparison for NamedSharding.

I'm less sure it would be a win to partially normalize PartitionSpec. Possibly normalizing length 1 tuples to strings instead would actually be more readable/less surprising.

**Comment by shoyer** [comment]
There is probably also room to improve the docs here to note this equivalence, e.g., on the PartitionSpec docstring :)
